[autoscaler] Fix local node provider

DmitriGekhtman opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

Closes https://github.com/ray-project/ray/issues/15342

This PR focuses on the most common use case for the local node provider: _bootstrapping ray at a given static set of ips_
This functionality was broken in the transition to available-node-type configs. This PR fixes it by
- Preparing local node provider configs with the available-node-type format for ingestion by autoscaler
- rsyncing the cluster state file so the head node is able to recognize itself as a non-terminated node
- modifying the node updater to avoid overriding node resources when using the local provider

Mostly to simplify testing, this PR also adds an optional `external_head_ip` so that you can `ray up` from a machine (e.g. your laptop) outside of the Ray cluster's network (e.g. an AWS VPC).

Default behavior is updated to be more intuitive, so for example users don't have to explicitly set min and max workers when specifying a list of ips.

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Added unit tests to
- Test bootstrapping of manually managed static clusters
- Check file-mounts are set to sync cluster state
- Check external head ip is processed correctly

Tested manually as follows
(recording for posterity if someone needs to debug this again):
`ray up` with the following AWS config
```
cluster_name: local-node-provider-test
max_workers: 3
upscaling_speed: 1.0
docker: {}
idle_timeout_minutes: 5
provider:
    type: aws
    region: us-west-2
    availability_zone: us-west-2a,us-west-2b
    cache_stopped_nodes: True
auth:
    ssh_user: ubuntu
    ssh_private_key: xxxx.pem
available_node_types:
    ray.head.default:
        min_workers: 0
        max_workers: 0
        resources: {}
        node_config:
            KeyName: xxxx
            InstanceType: m5.2xlarge
            ImageId: ami-0a2363a9cff180a64 # Deep Learning AMI (Ubuntu) Version 30
            BlockDeviceMappings:
                - DeviceName: /dev/sda1
                  Ebs:
                      VolumeSize: 100
    ray.worker.big:
        min_workers: 1
        max_workers: 1
        resources: {}
        node_config:
            KeyName: xxxx
            InstanceType: m5.xlarge
            ImageId: ami-0a2363a9cff180a64 # Deep Learning AMI (Ubuntu) Version 30
    ray.worker.default:
        min_workers: 1
        max_workers: 1
        resources: {}
        node_config:
            KeyName: xxxx
            InstanceType: m5.large
            ImageId: ami-0a2363a9cff180a64 # Deep Learning AMI (Ubuntu) Version 30
    ray.worker.gpu:
        min_workers: 1
        max_workers: 1
        resources: {}
        node_config:
            KeyName: xxxx
            InstanceType: p2.xlarge
            ImageId: ami-0a2363a9cff180a64 # Deep Learning AMI (Ubuntu) Version 30
head_node_type: ray.head.default
file_mounts: {}
cluster_synced_files: []
file_mounts_sync_continuously: False
rsync_exclude:
    - "**/.git"
    - "**/.git/**"
rsync_filter:
    - ".gitignore"
initialization_commands: []
setup_commands:
    - >-
        (stat $HOME/anaconda3/envs/tensorflow2_latest_p37/ &> /dev/null &&
        echo 'export PATH="$HOME/anaconda3/envs/tensorflow2_latest_p37/bin:$PATH"' >> ~/.bashrc) || true
    - which ray || pip install -U "ray[default] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl"
head_setup_commands:
    - pip install 'boto3>=1.4.8'  # 1.4.8 adds InstanceMarketOptions
worker_setup_commands: []
head_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --head --port=6379 --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml
worker_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --address=$RAY_HEAD_IP:6379 --object-manager-port=8076
head_node: {}
worker_nodes: {}
```
`ray attach` to the head node, start ipython, paste and run the following script to get aws node ids, instance types, resources, internal ips, external ips 
```
import yaml
import subprocess
from ray.autoscaler._private.aws.node_provider import AWSNodeProvider
config = yaml.safe_load(open("ray_bootstrap_config.yaml").read())
provider = AWSNodeProvider(config["provider"], config["cluster_name"])

nodes = provider.non_terminated_nodes({})

def get_type(node):
    return provider.node_tags(node)["ray-user-node-type"]

def node_type(node):
    type = get_type(node)
    return config["available_node_types"][type]

def node_config(node):
    type = get_type(node)
    return node_type(node)["node_config"]

def instance_type(node):
    return node_config(node)["InstanceType"]

def resources(node):
    return node_type(node)["resources"]

for node in nodes:
    print(node, instance_type(node), resources(node), "i"+provider.internal_ip(node), "e"+provider.external_ip(node))
```
In the same session, run the following to stop ray on all nodes
```
def runner(node):
    return provider.get_command_runner("dmitri",
                                        node,
                                        config["auth"],
                                        config["cluster_name"],
                                        subprocess,
                                        True,
                                        {})
def run(node, cmd):
    return runner(node).run(cmd, with_output=True)

print(subprocess.check_output("ray stop", shell=True))
for node in nodes[1:]:
    out = run(node, "ray stop")
    print(out)
```
Exit the ssh session.

`ray up` with the following "local node provider" config
```
cluster_name: minimal-manual

provider:
    type: local
    head_ip: <INTERNAL HEAD IP>
    external_head_ip: <EXTERNAL HEAD IP>
    worker_ips: [<INTERNAL WORKER IP 1>, <INTERNAL WORKER IP 2> , <INTERNAL WORKER IP 3>]

# How Ray will authenticate with newly launched nodes.
auth:
    ssh_user: ubuntu
    ssh_private_key: xxxx.pem

setup_commands:
    - if [ $(which ray) ]; then pip uninstall ray -y; fi
    - pip install -U "ray[default] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl"
    - if [ -d ray ]; then rm -Rf ray; fi
    - git clone --single-branch --branch fix-local-node-provider https://github.com/DmitriGekhtman/ray
    - pushd ray/python/ray && python setup-dev.py -y && popd
```
ray attach to the head node using the local config.
`ray status` shows the head and all workers with the correct resource counts:
```
======== Autoscaler status: 2021-06-02 21:22:13.680789 ========
Node status
---------------------------------------------------------------
Healthy:
 4 ray.node
Pending:
 (no pending nodes)
Recent failures:
 (no failures)

Resources
---------------------------------------------------------------

Usage:
 0.0/18.0 CPU
 0.0/1.0 GPU
 0.0/1.0 accelerator_type:K80
 0.00/75.396 GiB memory
 0.00/33.604 GiB object_store_memory

Demands:
 (no resource demands)
```
`ray down` to confirm that `ray stop` is executed successfully on the head node.

ijrsvt commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,94 @@
+    return config
+
+
+def is_local_manual(provider_config: Dict[str, Any]) -> bool:
+    return (provider_config["type"] == "local"
```
Quick doc string here, i.e. something like:
```
"Is this a 'LocalNodeProvider' that has manually specified IPs"
```

ijrsvt commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,94 @@
+
+def sync_state(config: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Used to synchronize head and local cluster state files, mostly to let the
+    head node know that the head node itself is non-terminated.
```
Can you update the docstring to make it clear that this doesn't copy files, but rather adds them as `file_mounts` that will be copied.

ijrsvt commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,94 @@
+    """
+    config = copy.deepcopy(config)
+    for field in "head_node", "worker_nodes", "available_node_types":
+        if config.get("head_node"):
```
```suggestion
        if config.get(field):
```

DmitriGekhtman replied:
Thanks, nice catch.

ijrsvt commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,94 @@
+    return config
+
+
+def prepare_manual(config: Dict[str, Any]) -> Dict[str, Any]:
+    config = copy.deepcopy(config)
+    if "worker_ips" not in config["provider"]:
+        cli_logger.error("Please supply a list of `worker_ips` "
```
Could we add a quick test for this?

DmitriGekhtman replied:
Added to testValidateDocker in test_autoscaler_yaml.py

ijrsvt commented on python/ray/autoscaler/local/defaults.yaml:
```diff
@@ -1,103 +1,50 @@
-# This executes all commands on all nodes in the docker container,
-# and opens all the necessary ports to support the Ray cluster.
-# Empty string means disabled. Assumes Docker is installed.
+# Docker use not recommended: ray down will not clean up Docker state.
```
Wait, `ray down` actually calls `docker stop` =)

DmitriGekhtman replied:
Right :) And I see docker run always gets an rm flag.
Let me get rid of the scary message..


For the docker stop to work properly in my experimental setup, I'd have to supply external ips for the workers... beh... (in theory not so bad to include another list) Maybe the external head ip setting was superfluous and I could have set up the experiment to run ray up from within the VPC. Will just leave that as is for now.

AmeerHajAli commented on python/ray/autoscaler/_private/commands.py:
```diff
@@ -691,6 +693,12 @@ def get_or_create_head_node(config: Dict[str, Any],
 
+        # If restarting Ray on a manually-managed on-prem cluster,
+        # we need to sync local and head representations of cluster state.
+        if ray_start_commands and is_local_manual(config["provider"]):
+            # Add cluster state file to file mounts.
+            config = sync_state(config)
```
not sure why ray_start_commands here are relevant.

DmitriGekhtman replied:
This isn't well-documented here and I'll add a comment explaining this:
The presence of `ray_start_commands` here indicates that you're restarting the cluster.
If you're not restarting the cluster, then doing the sync will break the autoscaler's state because the machine running ray up doesn't know about the worker nodes.

AmeerHajAli commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,98 @@
+        config = prepare_coordinator(config)
+    else:
+        config = prepare_manual(config)
+    # Presence of head_node and worker_nodes fields triggers legacy config
+    # processing later in the bootstrap process.
+    config["head_node"] = {"placeholder": {}}
+    config["worker_nodes"] = {"placeholder": {}}
```
Why are we doing it in this hacky way? Can we please not do this? didn't we discuss that we should not even need the legacy config processing? Just put everything under the 1 node type and it will work.

DmitriGekhtman replied:
Once I'm done with the rest of the changes in this PR, I'll try that out -- if it doesn't work out, the options will be likely that we keep it as is or that someone else from our team takes over.

IMO, attempting to use the autoscaler for the simple purpose of bootstrapping a Ray cluster on some private Ray clusters is pretty hacky as it is.

DmitriGekhtman replied:
Hmm, though the look is pretty hacky -- as a quick fix I'll refactor things to at least avoid having to put these placeholders in.

DmitriGekhtman replied:
Actually, not entirely sure how to remove that -- will try using a single node type later, but not entirely sure if the resource demand scheduler will be happy with that.

@AmeerHajAli 
My specific concern with using a single node type is that the resource demand scheduler may try to repeatedly infer runtime resources at every autoscaler update, because of this line:
https://github.com/ray-project/ray/blob/e6864523cf6c81e089cbbc41c50260b20ae23b42/python/ray/autoscaler/_private/resource_demand_scheduler.py#L311

AmeerHajAli replied:
I am not sure that this should be an issue, can you check it out?

DmitriGekhtman replied:
Using one node type appears to have worked fine and simplified the code.
Good call!

AmeerHajAli commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,98 @@
+    min_specified = "min_workers" in config
+    if not max_specified and not min_specified:
+        # The most common use-case.
+        config["min_workers"] = num_ips
```
is it ?

DmitriGekhtman replied:
From interactions with users, it seems the main use case for this NodeProvider is to bootstrap Ray at a fixed set of ips. (So, it's not really a cloud provider.) I don't think people are expecting autoscaling to work.

Also, neither the blog post on this nor the local node provider docs mention min_workers or max_workers.

Could modify the defaults for more consistency with the rest of the node providers, but I think it would be more confusing for users.

DmitriGekhtman replied:
Could be another reason to prefer min_workers=max_workers:
If I read the code right, it seems that downscaling appears to have never worked quite right for this node provider.

When a worker node is deleted, it is simply marked as [terminated](https://github.com/ray-project/ray/blob/6a227ae50132639552b7d3601cc828c28143e2cb/python/ray/autoscaler/_private/local/node_provider.py#L226) in the state file (`ray stop` isn't called). As future (low-priority) work we could try to fix that using, say a node_updater or command_runner in delete_node, but I don't think it's necessary right now.

AmeerHajAli commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,98 @@
+    num_ips = len(config["provider"]["worker_ips"])
+    max_specified = "max_workers" in config
+    min_specified = "min_workers" in config
+    if not max_specified and not min_specified:
```
why max specified or min _specified are in the same if statement? Can you please put each one alone?

AmeerHajAli commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,98 @@
+        # The most common use-case.
+        config["min_workers"] = num_ips
+        config["max_workers"] = num_ips
+    elif max_specified and not min_specified:
+        # Indicates intent to cap autoscaler usage.
+        # Infer conservative default.
+        config["min_workers"] = 0
+        cli_logger.warning("Inferring `min_workers:0`. Ray will start only on "
+                           "the head node. Ray may start on other nodes, "
+                           "according to workload.")
```
Why is this relevant? Lets keep things simple:
if not min_specified:
    config["min_workers"] = 0  # or num_ips based on Richard's response.
if not max_specified:
    config["max_workers"] = num_ips
What we have here is over complication/too much magic that will confused the end user.

DmitriGekhtman replied:
Sounds good.

DmitriGekhtman replied:
Fixed.

AmeerHajAli commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,98 @@
+
+
+def get_lock_path(cluster_name: str) -> str:
+    return "/tmp/cluster-{}.lock".format(cluster_name)
+
+
+def get_state_path(cluster_name: str) -> str:
+    return "/tmp/cluster-{}.state".format(cluster_name)
```
can you update everywhere else to use this? for example in the node provider to user get_*_path as well?

AmeerHajAli replied:
I saw you already do it below, this is great!

AmeerHajAli commented on python/ray/autoscaler/_private/local/node_provider.py:
```diff
@@ -196,7 +207,14 @@ def node_tags(self, node_id):
-        return socket.gethostbyname(node_id)
+        node_state = self.state.get()[node_id]
+        ext_ip = node_state.get("external_ip")
+        # If user has supplied an external ip, return that.
+        # (Useful when calling ray up from outside the network.)
```
I m not sure why we need this external_ip...

DmitriGekhtman replied:
I need it to run ray up from my laptop.

AmeerHajAli commented on python/ray/autoscaler/local/defaults.yaml:
```diff
@@ -1,103 +1,49 @@
-    - ray stop
-    - ray start --address=$RAY_HEAD_IP:6379
-
 head_node: {}
 worker_nodes: {}
```
remove?

DmitriGekhtman replied:
Done.

AmeerHajAli commented on python/ray/autoscaler/local/example-full.yaml:
```diff
@@ -49,11 +39,24 @@ auth:
-head_node: {}
+# The minimum number of worker nodes to add to the Ray cluster in addition to the head
+# node. This number should be >= 0.
+min_workers: MIN_NUMBER_OF_NODES_TO_KEEP_IN_THE_CLUSTER
```
can you set it to the default and say what is the default as it was previously?

DmitriGekhtman replied:
Will revert to [current](https://github.com/ray-project/ray/blob/59bf3a7b22b450f9c89caa16c6b46a0ebc6c4918/python/ray/autoscaler/local/example-full.yaml#L6-L14) state which is to set both to 0 and say what the typical desired behavior is.

DmitriGekhtman replied:
Done.

AmeerHajAli commented on python/ray/autoscaler/local/example-full.yaml:
```diff
@@ -49,11 +39,24 @@ auth:
+min_workers: MIN_NUMBER_OF_NODES_TO_KEEP_IN_THE_CLUSTER
+# The maximum number of workers nodes to add to the Ray cluster in addition to the head node.
+# This takes precedence over min_workers.
+max_workers: MAX_NUMBER_OF_NODES_TO_ADD_TO_THE_CLUSTER
```
same as previous comment.

DmitriGekhtman commented on python/ray/autoscaler/local/defaults.yaml:
```diff
@@ -1,103 +1,49 @@
+
+# The following block is used internally;
+# it does not represent user configuration.
 available_node_types:
```
Could you elaborate?

This is necessary for the cluster to be bootstrapped correctly.

DmitriGekhtman replied:
removed this from defaults -- configuring internally

DmitriGekhtman commented:
Addressed most of the above comments.
Remains:
(1) We should align on min_workers default
(2) I'm going to try to rejigger things to use a single node type.

AmeerHajAli commented:
Sounds good, @DmitriGekhtman , let me know if you need anything from my side!

DmitriGekhtman commented:
Fixing some weird failures in `test_autoscaler.py`.

DmitriGekhtman commented:
lol.. : https://github.com/ray-project/ray/blob/31364ed9b471980a20267eac6739aa8c1464fe94/python/ray/tests/test_autoscaler.py#L375

DmitriGekhtman commented:
Tests pass, please re-review.

To allow really leaving off "head_node" and "worker_nodes", I changed all instances of `config["head_node"]`and `config["worker_nodes"]` to `config.get("head_node", {})` and `config.get("worker_nodes", {})`.

DmitriGekhtman commented:
The single node type is now configured in `local.config.prepare_local`

ijrsvt commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,94 @@
+        cli_logger.abort("The field `max_workers` is required when using an "
+                         "automatically managed on-premise cluster.")
+    node_type = config["available_node_types"][NODE_TYPE]
+    node_type["min_workers"] = config.pop("min_workers", None) or 0
```
```suggestion
    node_type["min_workers"] = config.pop("min_workers", 0)
```

ijrsvt commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,94 @@
+        cli_logger.abort("The field `max_workers` is required when using an "
+                         "automatically managed on-premise cluster.")
+    node_type = config["available_node_types"][NODE_TYPE]
+    node_type["min_workers"] = config.pop("min_workers", 0)
```
```suggestion
	# Global `min_workers` are no longer used.
    node_type["min_workers"] = config.pop("min_workers", 0)
```

Or something like this!

ijrsvt reviewed (approved):
This looks good to me! Have you tested the most recent version?

DmitriGekhtman commented:
> This looks good to me! Have you tested the most recent version?

Yep -- did some final checks. Everything looks as expected with my experimental setup.

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -556,7 +556,7 @@ def launch_config_ok(self, node_id):
         node_type = node_tags.get(TAG_RAY_USER_NODE_TYPE)
 
-        launch_config = copy.deepcopy(self.config["worker_nodes"])
+        launch_config = copy.deepcopy(self.config.get("worker_nodes", {}))
```
can we completely ignore the worker_nodes field?

AmeerHajAli replied:
worth a comment

AmeerHajAli commented on python/ray/autoscaler/_private/commands.py:
```diff
@@ -597,7 +599,7 @@ def get_or_create_head_node(config: Dict[str, Any],
     cli_logger.newline()
     # TODO(ekl) this logic is duplicated in node_launcher.py (keep in sync)
-    head_node_config = copy.deepcopy(config["head_node"])
+    head_node_config = copy.deepcopy(config.get("head_node", {}))
```
worth a comment.

AmeerHajAli commented on python/ray/autoscaler/_private/local/config.py:
```diff
@@ -1,2 +1,101 @@
+unsupported_field_message = ("The field {} is not supported "
+                             "for on-premise clusters.")
+
+NODE_TYPE = "ray.node"
```
```suggestion
LOCAL_CLUSTER_NODE_TYPE = "local.cluster.node"
```

DmitriGekhtman commented:
did another check, just in case :)

AmeerHajAli commented on python/ray/autoscaler/_private/local/node_provider.py:
```diff
@@ -68,6 +70,13 @@ def __init__(self, lock_path, save_path, provider_config):
+                # Set external head ip, if provided by user.
+                # Useful if calling `ray up` from outside the network.
+                external_head_ip = provider_config.get("external_head_ip")
+                if external_head_ip:
```
This is still puzzling for me

DmitriGekhtman replied:
Glad to remove it, but I will no longer be able to test.

DmitriGekhtman replied:
This works exactly like external and internal ips for aws.

DmitriGekhtman replied:
(In fact that's what it is.)

DmitriGekhtman replied:
I'd refer to the AWS implementation for more context.

DmitriGekhtman replied:
Added a more detailed docstring and removed from example-minimal.yaml, as this is primarily a debugging tool.

DmitriGekhtman replied:
Elaborating a little on our setup -- this isn't something I was aware of previously:
The autoscaler tells its node updaters to use internal node ip:
https://github.com/ray-project/ray/blob/226754d92c14a321b5eb7efbb174817d6df1bf56/python/ray/autoscaler/_private/autoscaler.py#L623
The local cluster launcher does get the `use_internal_ip` arg and so uses an external_ip:
https://github.com/ray-project/ray/blob/226754d92c14a321b5eb7efbb174817d6df1bf56/python/ray/autoscaler/_private/autoscaler.py#L623

Whether or not to use an external or internal ip affects the ssh command here: 
https://github.com/ray-project/ray/blob/226754d92c14a321b5eb7efbb174817d6df1bf56/python/ray/autoscaler/_private/command_runner.py#L347

To be able to `ray up` from my laptop, I had to use an external ip.

DmitriGekhtman replied:
Thought about it -- using public ips everywhere in this context wouldn't work, as ray internals infer internal ips but autoscaler would try use the public ips, leading to key errors in ResourceDemandScheduler...

AmeerHajAli commented on python/ray/autoscaler/_private/commands.py:
```diff
@@ -691,6 +695,14 @@ def get_or_create_head_node(config: Dict[str, Any],
+        # we need to sync local and head representations of cluster state.
+        # If we're not restarting the cluster (empty ray start cmds), don't
+        # sync to avoid breaking on-prem cluster autoscaler state.
+        if ray_start_commands and is_local_manual(config["provider"]):
```
```suggestion
        if ray_start_commands and is_local_manual_node_provider(config["provider"]):
```

AmeerHajAli commented on python/ray/autoscaler/_private/commands.py:
```diff
@@ -691,6 +695,14 @@ def get_or_create_head_node(config: Dict[str, Any],
+        # If restarting Ray on a manually-managed on-prem cluster,
+        # we need to sync local and head representations of cluster state.
+        # If we're not restarting the cluster (empty ray start cmds), don't
+        # sync to avoid breaking on-prem cluster autoscaler state.
```
Why is this related to ray start commands though?

DmitriGekhtman replied:
ray start commands are non-empty if and only if the ray cluster is being restarted.
I'll make this clearer with variable naming.

DmitriGekhtman replied:
It now reads more like English:
`if restarting_ray and is_local_manual_node_provider`

DmitriGekhtman replied:
The ray up machine thinks that worker nodes are terminated because it never creates worker nodes. It know that the head node is non-terminated. The only time it is acceptable to sync up that state is when ray is being started or restarted.

DmitriGekhtman commented:
Reviewers can merge at their discretion.

AmeerHajAli commented on python/ray/autoscaler/_private/local/node_provider.py:
```diff
@@ -196,7 +208,21 @@ def node_tags(self, node_id):
+
+        This can be used to call ray up from outside the network, for example
+        if the Ray cluster exists in an AWS VPC and we're interacting with
+        the cluster from a laptop.
```
```suggestion
        the cluster from a laptop (where using an internal_ip will not work).
```

AmeerHajAli commented on python/ray/autoscaler/_private/local/node_provider.py:
```diff
@@ -68,6 +70,14 @@ def __init__(self, lock_path, save_path, provider_config):
 
+                # Set external head ip, if provided by user.
+                # Necessary if calling `ray up` from outside the network.
+                # Refer to LocalNodeProvider.external_ip.
```
```suggestion
                # Refer to LocalNodeProvider.external_ip function.
```

AmeerHajAli commented on python/ray/autoscaler/_private/commands.py:
```diff
@@ -691,6 +695,16 @@ def get_or_create_head_node(config: Dict[str, Any],
             ray_start_commands = config["head_start_ray_commands"]
 
+        # If restarting Ray on a manually-managed on-prem cluster,
+        # we need to sync local and head representations of cluster state.
```
```suggestion
        # we need to upload the local cluster state file to the head node.
```

AmeerHajAli commented on python/ray/autoscaler/_private/commands.py:
```diff
@@ -691,6 +695,16 @@ def get_or_create_head_node(config: Dict[str, Any],
+        # If restarting Ray on a manually-managed on-prem cluster,
+        # we need to sync local and head representations of cluster state.
+        # If we're not restarting the cluster (empty ray start cmds), don't
+        # sync to avoid breaking on-prem cluster autoscaler state.
```
```suggestion
        # sync to avoid breaking the currently running on-prem cluster autoscaler state.
```

AmeerHajAli commented on python/ray/autoscaler/_private/util.py:
```diff
@@ -143,6 +144,10 @@ def prepare_config(config: Dict[str, Any]) -> Dict[str, Any]:
     - Has a valid Docker configuration if provided.
     - Has max_worker set for each node type.
     """
+    is_local = config["provider"]["type"] == "local"
```
can this raise a `KeyError`?

DmitriGekhtman replied:
thanks for fixing -- looking forward to the day oss code uses pydantic instead of nested dicts

AmeerHajAli commented on python/ray/autoscaler/_private/util.py:
```diff
@@ -143,6 +144,10 @@ def prepare_config(config: Dict[str, Any]) -> Dict[str, Any]:
     - Has a valid Docker configuration if provided.
     - Has max_worker set for each node type.
     """
+    is_local = config["provider"]["type"] == "local"
```
```suggestion
    is_local = config.get("provider", {}).get("type") == "local"
```
